oeps/oep-0006.rst

```diff
@@ -0,0 +1,410 @@
+
+  * Example: The student's answer to a multiple choice question
+* ``Scope.user_state_summary`` (UserScope.ALL, BlockScope.USAGE)
+
+  * Example: The number of students who have answered the question (side note:
+    there is no locking or atomicity for XBlock field writes, so in practice
+    this field is currently of little use in a MOOC)
+* ``Scope.user_info`` (UserScope.ONE, BlockScope.ALL)
+
+  * Example: user age (note: this scope seems little-used and overlaps with the
+    user service)
+* ``Scope.preferences`` (UserScope.ONE, BlockScope.TYPE)
+
+  * Example: (For a video XBlock) Boolean setting to prefer YouTube or
+    non-YouTube video playback.
+
+
+
+Motivation
+==========
+
+As one would expect, ``Scope.user_info`` and ``Scope.preferences`` fields in the
+Open edX platform are implemented globally, and not scoped to the course that
+uses them. There is currently no standard, extensible way to use XBlock fields
+to store data that is scoped to a particular "context," such as a course,
+although this is a common need.
+
+In the absence of a formal method for achieving context-scoped fields through
+the XBlock API, many other solutions have been put in place, such as: adding
+custom Django models to XBlocks, storing fields on the root "course" XBlock,
+defining inherited fields using InheritanceMixin, using edx-submissions as a
+generic data store, and more.
+
+It is hoped that the new canonical mechanism described by this OEP can provide
+consistency, improve interoperability, allow new types of behavior to be
+implemented by XBlocks, and give both authors and learners additional control
+over learner interactions with XBlocks.
+
+
+
+Specification
+=============
+
+New BlockScope value: CONTEXT
+-----------------------------
+
+To support context-scoped fields, ``BlockScope`` will allow a new value:
+``CONTEXT``. This new ``BlockScope`` will denote data that is related to all
+instances of this type of XBlock within a particular context.
+
+A "context" is defined by the runtime but is usually a "course" or the nearest
+equivalent. In the Open edX platform, the context may be a course or a content
+library. Different runs of the same course would be considered different
+contexts and would not share data, though UserScope.NONE fields would be copied
+when creating a new course run. On other platforms, a "context" could be defined
+as a blog post or a textbook.
+
+New named scopes
+----------------
+
+As is the situtation today, XBlock runtimes would generally not be expected to
+implement every possible combination of ``BlockScope`` and ``UserScope``.
+Instead, XBlock runtimes are encouraged to implement these new named scopes,
+which would be defined in the XBlock python package:
+
+* ``Scope.context_settings`` (UserScope.NONE, BlockScope.CONTEXT)
+
+  * Example: For an LTI XBlock, a list of the LTI provider URLs and secrets used
+    in the course
+  * Example: For a discussions XBlock: Enable or disable discussions in this
```

Comment:
Yeah, this OEP would not provide as much value for the discussions use case if we are not planning to use a `discussion` XBlock that supports both inline and course-tab views, and it would not support the other contexts that @nasthagiri mentions there